src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value_impl.hpp

```diff
@@ -0,0 +1,457 @@
+/**
+ * @file discrete_hilbert_value.hpp
+ * @author Mikhail Lozhnikov
+ *
+ * Defintion of the DiscreteHilbertValue class, a class that calculates
+ * the ordering of points using the Hilbert curve.
+ */
+#ifndef MLPACK_CORE_TREE_RECTANGLE_TREE_DISCRETE_HILBERT_VALUE_IMPL_HPP
+#define MLPACK_CORE_TREE_RECTANGLE_TREE_DISCRETE_HILBERT_VALUE_IMPL_HPP
+
+#include "discrete_hilbert_value.hpp"
+
+namespace mlpack {
+namespace tree /** Trees and tree-building procedures. */ {
+
+template<typename TreeElemType>
+DiscreteHilbertValue<TreeElemType>::DiscreteHilbertValue() :
+    localDataset(NULL),
+    ownsLocalDataset(false),
+    numValues(0),
+    valueToInsert(NULL),
+    ownsValueToInsert(false)
+{
+
+}
+
+template<typename TreeElemType>
+DiscreteHilbertValue<TreeElemType>::~DiscreteHilbertValue()
+{
+  if (ownsLocalDataset)
+    delete localDataset;
+  if (ownsValueToInsert)
+    delete valueToInsert;
+}
+
+template<typename TreeElemType>
+template<typename TreeType>
+DiscreteHilbertValue<TreeElemType>::DiscreteHilbertValue(const TreeType* tree) :
+    localDataset(NULL),
+    ownsLocalDataset(false),
+    numValues(0),
+    valueToInsert(tree->Parent() ? 
+                tree->Parent()->AuxiliaryInfo().HilbertValue().ValueToInsert() :
+                new arma::Col<HilbertElemType>(tree->Dataset().n_rows)),
+    ownsValueToInsert(tree->Parent() ? false : true)
+{
+  // Calculate the Hilbert value for all points
+  if (!tree->Parent()) //  This is the root node
+    ownsLocalDataset = true;
+  else if (tree->Parent()->Children()[0]->IsLeaf())
+  {
+    // This is a leaf node
+    assert(tree->Parent()->NumChildren() > 0);
+    ownsLocalDataset = true;
+  }
```

Comment:
Should we change this now?  If we have decided that the `localDataset` should be removed from `RectangleTree`, then I think that we can change `localDataset` and `ownsLocalDataset` here to `dataset` and `ownsDataset`, and make `ownsDataset` only true when the node is the root, not also when it is a leaf.  Maybe `hilbertValues` is a better name than `dataset`, since that object is holding the Hilbert values for each dimension, not the actual dataset itself.